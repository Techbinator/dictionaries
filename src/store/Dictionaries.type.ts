export interface IDictionary {
  uuid: string;
  name: string;
  translationsUUID: string[];
}

export interface ITranslation {
  uuid: string;
  domain: string;
  range: string;
}

export interface IDictionaries {
  [uuid: string]: IDictionary;
}

export interface ITranslations {
  [uuid: string]: ITranslation;
}
export type TeditTranslation = (translation: ITranslation) => void;
export type TaddTranslation = (
  dictionaryUUID: string,
  translation: ITranslation
) => void;
export type TremoveTranslation = (
  dictionaryUUID: string,
  translationUUID: string
) => void;
export type TaddDictionary = (name: string) => void;
export type TremoveDictionary = (uuid: string) => void;
export type TactivateDictionary = (uuid: string) => void;
export interface IDictionaryProvider {
  dictionaries: IDictionaries;
  translations: ITranslations;
  addDictionary: TaddDictionary;
  activeDictionary: string;
  removeDictionary: TremoveDictionary;
  removeTranslation: TremoveTranslation;
  addTranslation: TaddTranslation;
  editTranslation: TeditTranslation;
  activateDictionary: TactivateDictionary;
}
