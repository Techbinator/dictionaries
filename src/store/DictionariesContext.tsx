import React from "react";
import {
  IDictionaryProvider,
  IDictionaries,
  ITranslation,
  ITranslations
} from "./Dictionaries.type";
import { v1 } from "uuid";

export const DictionariesContext = React.createContext({});
export class DictionariesProvider extends React.Component {
  addDictionary = (name: string) => {
    const uuid = v1();
    const newDictionary = {
      uuid,
      name,
      translationsUUID: []
    };
    this.setState({
      dictionaries: { [uuid]: newDictionary, ...this.state.dictionaries }
    });
  };

  activateDictionary = (dictionaryUUID: string) => {
    this.setState({ activeDictionary: dictionaryUUID });
  };

  removeDictionary = (uuid: string) => {
    const { activeDictionary, dictionaries, translations } = this.state;
    const dictionary = dictionaries[uuid];
    const newDictionaries: IDictionaries = {
      ...dictionaries
    };
    const NewTranslations = {
      ...translations
    };
    dictionary.translationsUUID.forEach(uuid => delete NewTranslations[uuid]);

    delete newDictionaries[uuid];
    this.setState({
      dictionaries: newDictionaries,
      activeDictionary: activeDictionary === uuid ? "" : activeDictionary,
      translations: NewTranslations
    });
  };

  removeTranslation = (dictionaryUUID: string, translationUUID: string) => {
    const { dictionaries, translations } = this.state;
    const newTranslations: ITranslations = {
      ...translations
    };
    delete newTranslations[translationUUID];
    this.setState({
      translations: newTranslations,
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaries[dictionaryUUID],
          translationsUUID: dictionaries[
            dictionaryUUID
          ].translationsUUID.filter(tuuid => tuuid !== translationUUID)
        }
      }
    });
  };

  addTranslation = (dictionaryUUID: string, translation: ITranslation) => {
    const { dictionaries, translations } = this.state;
    const uuid = v1();
    const newTranslation = {
      ...translation,
      uuid
    };

    const dictionaryTranslations = dictionaries[dictionaryUUID];
    this.setState({
      translations: {
        [uuid]: newTranslation,
        ...translations
      },
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaryTranslations,
          translationsUUID: [uuid, ...dictionaryTranslations.translationsUUID]
        }
      }
    });
  };

  editTranslation = (translation: ITranslation) => {
    this.setState({
      translations: {
        ...this.state.translations,
        [translation.uuid]: translation
      }
    });
  };

  state: IDictionaryProvider = {
    dictionaries: {},
    translations: {},
    activeDictionary: "",
    addDictionary: this.addDictionary,
    removeDictionary: this.removeDictionary,
    removeTranslation: this.removeTranslation,
    addTranslation: this.addTranslation,
    editTranslation: this.editTranslation,
    activateDictionary: this.activateDictionary
  };

  //get from localstorage
  componentWillMount() {
    const dictionaries = localStorage.getItem("dictionaries");
    if (dictionaries) {
      this.setState({ dictionaries: JSON.parse(dictionaries) });
    }

    const translations = localStorage.getItem("translations");
    if (translations) {
      this.setState({ translations: JSON.parse(translations) });
    }
  }
  //save to localstorage
  componentDidUpdate({}, prevState: IDictionaryProvider) {
    const { dictionaries, translations } = this.state;
    if (
      JSON.stringify(dictionaries) !== JSON.stringify(prevState.dictionaries)
    ) {
      // save to localstorage
      localStorage.setItem("dictionaries", JSON.stringify(dictionaries));
    }
    if (
      JSON.stringify(translations) !== JSON.stringify(prevState.translations)
    ) {
      // save to localstorage
      localStorage.setItem("translations", JSON.stringify(translations));
    }
  }

  render() {
    return (
      <DictionariesContext.Provider value={this.state}>
        {this.props.children}
      </DictionariesContext.Provider>
    );
  }
}
