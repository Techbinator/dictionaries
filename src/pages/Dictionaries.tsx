import React from "react";

import Dictionary from "../components/Dictionary";
import NewDictionary from "../components/NewDictionary";
import { DictionariesContext } from "../store/DictionariesContext";

export default class Dictionaries extends React.Component {
  render() {
    const {
      dictionaries,
      translations,
      addDictionary,
      addTranslation,
      removeDictionary,
      removeTranslation,
      editTranslation
    } = this.context;
    return (
      <table className="one-edge-shadow">
        <thead>
          <tr>
            <th colSpan={2}>Name</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          <NewDictionary addDictionary={addDictionary} />
          {Object.keys(dictionaries).map((keyName: string) => (
            <Dictionary
              key={dictionaries[keyName].uuid}
              dictionary={dictionaries[keyName]}
              translations={translations}
              addTranslation={addTranslation}
              removeDictionary={removeDictionary}
              removeTranslation={removeTranslation}
              editTranslation={editTranslation}
            />
          ))}
        </tbody>
      </table>
    );
  }
}
Dictionaries.contextType = DictionariesContext;
