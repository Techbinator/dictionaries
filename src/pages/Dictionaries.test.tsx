import React from "react";
import Dictionaries from "./Dictionaries";
import { mount } from "enzyme";
import { DictionariesProvider } from "../store/DictionariesContext";
import contextMock from "../helpers/__mocks__/context.mock";

describe("Dictionaries component", () => {
  it("should render snapshot with default values", () => {
    const wrapper = mount(
      <DictionariesProvider {...contextMock}>
        <Dictionaries />
      </DictionariesProvider>,
      { context: contextMock }
    );
    expect(wrapper).toMatchSnapshot();
  });
});
