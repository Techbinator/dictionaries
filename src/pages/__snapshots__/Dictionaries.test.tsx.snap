// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionaries component should render snapshot with default values 1`] = `
<DictionariesProvider
  activateDictionary={[MockFunction]}
  activeDictionary="619ce930-4116-11e9-9306-ef359c18301d"
  addDictionary={[MockFunction]}
  addTranslation={[MockFunction]}
  dictionaries={
    Object {
      "4aa2cce0-4116-11e9-9306-ef359c18301d": Object {
        "name": "test",
        "translationsUUID": Array [
          "58f6b540-4116-11e9-9306-ef359c18301d",
          "55a62ab0-4116-11e9-9306-ef359c18301d",
        ],
        "uuid": "4aa2cce0-4116-11e9-9306-ef359c18301d",
      },
      "619ce930-4116-11e9-9306-ef359c18301d": Object {
        "name": "dictionary2",
        "translationsUUID": Array [
          "6e270e10-4116-11e9-9306-ef359c18301d",
        ],
        "uuid": "619ce930-4116-11e9-9306-ef359c18301d",
      },
    }
  }
  editTranslation={[MockFunction]}
  removeDictionary={[MockFunction]}
  removeTranslation={[MockFunction]}
  translations={
    Object {
      "55a62ab0-4116-11e9-9306-ef359c18301d": Object {
        "domain": "Crimson",
        "range": "newText",
        "uuid": "55a62ab0-4116-11e9-9306-ef359c18301d",
      },
      "58f6b540-4116-11e9-9306-ef359c18301d": Object {
        "domain": "a",
        "range": "b",
        "uuid": "58f6b540-4116-11e9-9306-ef359c18301d",
      },
      "6e270e10-4116-11e9-9306-ef359c18301d": Object {
        "domain": "Turquoise",
        "range": "newRange",
        "uuid": "6e270e10-4116-11e9-9306-ef359c18301d",
      },
    }
  }
>
  <Dictionaries>
    <table
      className="one-edge-shadow"
    >
      <thead>
        <tr>
          <th
            colSpan={2}
          >
            Name
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <NewDictionary
          addDictionary={[Function]}
        >
          <tr>
            <td
              className="p-relative"
              colSpan={2}
            >
              <input
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Name"
                type="string"
                value=""
              />
            </td>
            <td>
              <input
                onClick={[Function]}
                type="button"
                value="Add dictionary"
              />
            </td>
          </tr>
        </NewDictionary>
      </tbody>
    </table>
  </Dictionaries>
</DictionariesProvider>
`;
